dev/PushNotifications/PushNotificationChannel.cpp

```diff
@@ -32,141 +31,89 @@ namespace PushNotificationHelpers
-                }
+            auto lock { m_lock.lock_shared() };
+            registeredEvent = bool(m_foregroundHandlers);
+        }
 
-                ++m_foregroundHandlerCount;
+        if(!registeredEvent)
+        {
+            if(PushNotificationHelpers::IsPackagedAppScenario())
+            {
+                auto appUserModelId{ winrt::Microsoft::Helpers::GetAppUserModelId() };
 
-                return m_foregroundHandlers.add(handler);
+                // Register foreground event through the UDK
+                THROW_IF_FAILED(PushNotifications_RegisterNotificationSinkForFullTrustApplication(appUserModelId.get(), this));
             }
-        }
-        else
-        {
-            auto lock = m_lock.lock_exclusive();
-            if (!m_foregroundHandlerCount++)
+            else
             {
                 auto notificationsLongRunningPlatform{ winrt::Microsoft::Helpers::GetNotificationPlatform() };
 
                 wil::unique_cotaskmem_string processName;
                 THROW_IF_FAILED(GetCurrentProcessPath(processName));
 
-                THROW_IF_FAILED(notificationsLongRunningPlatform->RegisterForegroundActivator(this, processName.get()));             
+                // Register foreground event through the LRP
+                THROW_IF_FAILED(notificationsLongRunningPlatform->RegisterForegroundActivator(this, processName.get()));
             }
-            return m_foregroundHandlers.add(handler);
         }
+
+        auto lock { m_lock.lock_exclusive() };
+        return m_foregroundHandlers.add(handler);
     }
 
     void PushNotificationChannel::PushReceived(winrt::event_token const& token) noexcept
     {
-        if (PushNotificationHelpers::IsPackagedAppScenario())
+        bool registeredEvent { false };
         {
-            if (m_channel)
-            {
-                m_channel.PushNotificationReceived(token);
-            }
-            else
-            {
-                auto lock = m_lock.lock_exclusive();
-
-                if (m_foregroundHandlerCount == 1)
-                {
-                    auto appUserModelId{ winrt::Microsoft::Helpers::GetAppUserModelId() };
+            auto lock { m_lock.lock_exclusive() };
+            m_foregroundHandlers.remove(token);
+            registeredEvent = bool(m_foregroundHandlers);
+        }
 
-                    THROW_IF_FAILED(PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get()));
-                }
+        if(!registeredEvent)
+        {
+            if (PushNotificationHelpers::IsPackagedAppScenario())
+            {
+                auto appUserModelId{ winrt::Microsoft::Helpers::GetAppUserModelId() };
 
-                m_foregroundHandlers.remove(token);
-                --m_foregroundHandlerCount;
+                // Unregister foreground sink in UDK
+                THROW_IF_FAILED(PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get()));
```

Comment:
You cannot remove the foreground sink unless there are no registered handlers.

Reply:
!registeredEvent checks if there are no handlers registered

Reply:
Comment resolved above